1.12 Overhaul (Grid Generation, Docs, Bugfixes)

SizableShrimp opened the pull request:
This is a backwards-compatible breaking change to how the 1.12.2 (Compact Machines 3) mod works. I have bumped the version to 3.1.0 to reflect this. Depending on what is decided to do with this PR, it could be the final 1.12.2 update before 1.16.4 is released.

I reviewed all the code I modified and have tested it in-game, but redundancy for testing is always good. I have been running this in production with a small server without issues, and the issues that have arose, I have fixed. There are some more recent changes I made in the last 24 hours that I haven't yet gotten to test specifically, but they are mainly just code refactors that don't change much actual logic of how the mod works. Here is a link to a compiled JAR for testing purposes: [Compact Machines 3.1.0](https://cdn.discordapp.com/attachments/718807454703550544/793217992090451998/compactmachines3-1.12.2-3.1.0.jar)

### Summary of changes
- **Add 2d grid generation**
- Support older 3.0.18 file formats
- Update build.gradle
- Manually include RS API
- Add /compactmachines3 documentation command
- Minor bug fixes
- Translation overhaul
- Logger refactoring
- and other stuff I have missed

A majority of the code refactoring was switching code dependent on `mod 1024` to its appropriate `roomPos` equivalent by getting the room position data from the `WorldSavedDataMachines`. The `roomPos` is the very northwest corner of the room at y=40. It is the corner of the machine wall box. The first one would generate at (0, 40, 0), and then it would populate the grid around (0, 0). The `roomPos` generation is backwards-compatible so that it doesn't overwrite old machine rooms. All changed NBT formats are also backwards-compatible in every single way that I could make possible. This means that any old save data on 3.0.18 will be seamlessly transferred to the newer format. Note that **this NBT data update is not reversible and save data is permanently altered**. This should be fine for most people, but "trying out" the newer mod version and then changing the mod back if you don't like it is not an option *unless* a backup is made. For the majority of users and modpack creators that are still on 1.12.2, this update should not negatively affect anything and should be a great plus for server owners (reasoning explained below). The room position of each room is displayed inside of the GUI of machine previews if users are admins. This allows for administrators to easily navigate to rooms if they need. The relevant teleportation commands still function and teleport admins to the correct room location. One weak point of CM3 is the poor command documentation and overall command structure. For example, it is not stated anywhere that doing `/compactmachines3 machines view (insert machine id here)` will display the specific machine id. This could be something to improve in a further PR.

## Reasoning for the change
For large servers, having ALL compact machines on *one dimensional axis* is very inefficient on server resources and fosters a bad user-graphical experience. Since every machine is generate in the form `(1024*machine id, 40, 0)`, this means that after only 499 generated machines, the 500th machine is located at `(512000, 40, 0)`. At this distance and further, many rendering bugs with Tile Entities and graphical mod displays appear. For example, the mod's own miniaturization field projectors sweeping monitor motion completely bugs out because of **floating-point rounding precision errors** that appear at large distances. The 1.12.2 Minecraft rendering engine (LWJGL 2) simply does not perform well with rendering Tile Entities and other things at large distances away from `(0, 0)`. By converting generation to a 2-dimensional system (and supporting backwards-compatibility for already-generated machines), immediate improvements can be seen on servers with many machines. All machines generated after updating from 3.0.18 to 3.1.0 are now generated in a 2-dimensional fashion starting from the origin. The format for a machine generated on 3.1.0 would now be `(512*x, 40, 512*y)` where `x` and `y` are dependent on the generation code and any existing older rooms. Any older machines would be converted to this new `roomPos` storage format as well. This means that any newly generated machines will immediately lose all rendering bugs and issues that appear on 3.0.18. The same 500 machines that generated out to `(512000, 40, 0)` in 3.0.18 could fit in an approximately 11,000 x 11,000 square in 3.1.0. This is nowhere near the amount of distance needed with only one dimension. These gains also improve with more generated machines since the perimeter of the square increases faster than its side length. This is my reasoning for making these large changes.

I hope this PR can be tested and accepted before the 1.16.4 version (Compact Machines 4) is released. I'd hate to see this work wasted when it brings so much improvements for 1.12.2 users and modpacks that still might want to incorporate the mod.

SizableShrimp committed:
Add 2-dimensional grid support to machine generation. Fixes #429

This change was designed with backwards compatability in mind, but is not API compatible.
This means the version has been bumped to 3.1.0 in line with semantic versioning.
This new version **MUST** be the same on both client and server. It *might* be possible to not have this on the client, but it is unsupported.

Summary of changes:
Add 2d grid generation
Support older 3.0.18 file formats
Update build.gradle
Manually include RS API
Add /compactmachines3 documentation command
Minor bug fixes
Code formatting
Translation overhaul
Logger refactoring
and other stuff I have missed

SizableShrimp committed:
Fix #422 and #426

The changes to how an instance of WorldSavedDataMachines is created mostly fixed this, but I added redudancy.

SizableShrimp committed:
Fix compile-time errors

robotgryphon commented on src/main/java/org/dave/compactmachines3/CompactMachines3.java:
```diff
@@ -7,39 +7,54 @@
+    public static final Logger logger = LogManager.getLogger(CompactMachines3.MODID);
 
     public static ClientWorldData clientWorldData;
+    public static Map<Integer, BlockPos> clientMachineGrid;
+    public static Map<Integer, EnumMachineSize> clientMachineSizes;
```
This seems a bit odd placed here; done in a WSD instance on the 1.16 version. Any particular reasoning?

SizableShrimp replied:
No particular reason; I mainly did it since there is another client-side field that used the main mod class for this even though it is quite improper. I can move this to the WSD class once I get home. Collaborators also have access to push commits to my fork if you care enough to do it yourself.

SizableShrimp replied:
Fixed

robotgryphon commented on src/main/java/org/dave/compactmachines3/block/BlockMachine.java:
```diff
@@ -227,15 +228,20 @@ public static void spawnItemWithNBT(World world, BlockPos pos, EnumMachineSize s
-            compound.setUniqueId("owner", te.getOwner());
-        }
-        if(te.hasNewSchema()) {
-            compound.setString("schema", te.getSchemaName());
+        // If a player places a machine, doesn't use it, and picks it back up, they can't stack it as it has NBT data.
+        if (ConfigurationHandler.MachineSettings.allowPickupEmptyMachines && te.id == -1 && !te.hasNewSchema()) {
+            // Don't add NBT data if the config value is set to true and the machine id is -1 and it doesn't have a schema
+        } else {
+            NBTTagCompound compound = new NBTTagCompound();
+            compound.setInteger("machineId", te.id);
+            if (te.hasOwner()) {
+                compound.setUniqueId("owner", te.getOwner());
+            }
+            if (te.hasNewSchema()) {
+                compound.setString("schema", te.getSchemaName());
+            }
+            stack.setTagCompound(compound);
```
Possibly breaking change, did this config value exist before, or was it not documented?

SizableShrimp replied:
It did not exist before. I added it, but it defaults to false so as to explicitly not cause breaking changes. I refactored the config a bit, mainly adding more language key translations and some refactoring. You can find these changes in the `ConfigurationHandler` class I believe.

robotgryphon commented on src/main/java/org/dave/compactmachines3/block/BlockRedstoneTunnel.java:
```diff
@@ -81,16 +81,19 @@ public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, En
-            while(sideMapping != null) {
+            int id = StructureTools.getIdForPos(pos);
+            Map<EnumFacing, RedstoneTunnelData> sideMapping = WorldSavedDataMachines.getInstance().redstoneTunnels.get(id);
+            if (sideMapping == null) {
+                return true;
+            }
+            while (true) {
```
Single null check over a long-running while loop, looks like?

SizableShrimp replied:
This loop is weird. I didn't investigate it much, but IntelliJ notified me that the `sideMapping` can only be null once since it is not updated inside the loop. I'm not sure what the loop is doing, but I'd rather not change something I don't understand.

robotgryphon commented on src/main/java/org/dave/compactmachines3/command/CommandSchemaLoad.java:
```diff
@@ -43,14 +43,14 @@ public void execute(MinecraftServer server, ICommandSender sender, String[] args
-        EnumMachineSize machineSize = WorldSavedDataMachines.INSTANCE.machineSizes.get(coords);
-        if(machineSize != schema.getSize()) {
-            throw this.getException(sender, "machine_size_does_not_match");
+        int id = WorldSavedDataMachines.getInstance().getMachineIdFromBoxPos(sender.getCommandSenderEntity());
```
I understand the intent from context, but `BoxPos` on a `PlayerEntity` instance is a bit misleading here

SizableShrimp replied:
I think what I was going for with `BoxPos` was "a position inside the box", which an entity could be considered a position. If you have a better name for this overloaded method, feel free to tell me.

SizableShrimp replied:
I have updated this method to be called `getMachineIdFromEntityPos` for the entity overload method.

robotgryphon commented on src/main/java/org/dave/compactmachines3/gui/framework/widgets/Widget.java:
```diff
@@ -115,10 +114,10 @@ public String toString() {
     public boolean isPosInside(int x, int y) {
         boolean isInsideX = this.getActualX() <= x && x < this.getActualX() + this.width;
-        //Logz.info("[%s] insideX: %d < %d <= %d --> %s", this.id, this.lastDrawX, x, this.lastDrawX + this.lastDrawWidth, this.toString());
+        //Logz.info("[{}] insideX: {} < {} <= {} --> {}", this.id, this.lastDrawX, x, this.lastDrawX + this.lastDrawWidth, this.toString());
```
If these are commented out, better to remove the lines since they haven't been used in production for a while

SizableShrimp replied:
Sure. I mainly left them since they were there before. I was using a search and replace function to replace al the string formats with logger formats. I can remove them when I get home.

robotgryphon commented on src/main/java/org/dave/compactmachines3/gui/framework/widgets/WidgetButton.java:
```diff
@@ -63,7 +63,7 @@ public void setUnlocalizedLabel(String unlocalizedLabel) {
     @Override
     public void draw(GuiScreen screen) {
-        //Logz.info("Width: %d, height: %d", width, height);
+        //Logz.info("Width: {}, height: {}", width, height);
```
Prefer full removal of unnecessary logging

SizableShrimp replied:
Noted

robotgryphon commented on src/main/java/org/dave/compactmachines3/gui/framework/widgets/WidgetInputField.java:
```diff
@@ -50,7 +50,7 @@ public WidgetInputField(String fieldIdentifier) {
             @Override
             public WidgetEventResult call(MouseClickEvent event, Widget widget) {
-                //Logz.info("Mouse clicked @ x=%d, y=%d", event.x, event.y);
+                //Logz.info("Mouse clicked @ x={}, y={}", event.x, event.y);
```
Prefer full removal of unnecessary logging

SizableShrimp replied:
Noted

robotgryphon commented on src/main/java/org/dave/compactmachines3/gui/framework/widgets/WidgetList.java:
```diff
@@ -71,7 +71,7 @@ public void draw(GuiScreen screen) {
         Gui.drawRect(1, 1, width-1, height-1, backgroundColor);
 
-        //Logz.info("Rendering lines %d to %d", lineOffset, lastVisibleLine);
+        //Logz.info("Rendering lines {} to {}", lineOffset, lastVisibleLine);
```
Prefer full removal of unnecessary logging

SizableShrimp replied:
Noted

robotgryphon commented on src/main/java/org/dave/compactmachines3/gui/framework/widgets/WidgetPanel.java:
```diff
@@ -21,7 +21,7 @@ public WidgetPanel() {
         this.addListener(MouseMoveEvent.class, (event, widget)-> {
 
-            //Logz.info("[%s] Moved mouse to x=%d, y=%d", widget.id, innerX, innerY);
+            //Logz.info("[{}] Moved mouse to x={}, y={}", widget.id, innerX, innerY);
```
Prefer full removal of unnecessary logging

SizableShrimp replied:
Noted

robotgryphon commented on src/main/java/org/dave/compactmachines3/gui/psd/Page.java:
```diff
@@ -33,10 +35,10 @@ public void draw(GuiPSDScreen guiPSDScreen, FontRenderer fontRenderer, RenderIte
 
     public void mouseClicked(GuiPSDScreen psd, int mouseX, int mouseY, int mouseButton) {
         if(mouseButton == 1) {
-            if(name == "welcome") {
+            if(name.equals("welcome")) {
```
Should probably leave the `==` check here, and maybe consider moving the value to a const somewhere..

SizableShrimp replied:
`==` check?? It's a String....

robotgryphon commented on src/main/java/org/dave/compactmachines3/miniaturization/MultiblockRecipes.java:
```diff
@@ -46,9 +45,10 @@ public static MultiblockRecipe tryCrafting(World world, BlockPos projectorPos, I
-        if(opt.isPresent()) {
-            return opt.get();
+        for (MultiblockRecipe recipe : recipes) {
+            if (recipe.getName().equals(name)) {
+                return recipe;
+            }
```
Performance change?

SizableShrimp replied:
? Streams are slower than their imperative counterpart.

robotgryphon commented on src/main/java/org/dave/compactmachines3/miniaturization/MultiblockRecipe.java:
```diff
@@ -159,7 +160,7 @@ public void setPositionMap(String[][][] map) {
                     }
 
-                    referenceCount.put(content, referenceCount.getOrDefault(content, 0)+1);
+                    referenceCount.merge(content, 1, Integer::sum);
```
Assuming this is an automated update from Java 8?

SizableShrimp replied:
Not automated, no. I manually modified this once I saw it was not using the proper solution since I was in that file for some other reason and it caught my eye.

SizableShrimp committed:
Remove unused logs

SizableShrimp committed:
Refactor `clientMachineGrid` and `clientMachineSizes`

Moved the fields to `WorldSavedDataMachines`

SizableShrimp committed:
Fix NPE when world is loading and roomPos has not yet loaded

SizableShrimp committed:
Allow operators to bypass locked machines. Closes #399

Added documentation to `StructureTools#getInsetWallFacing` and updated signature

robotgryphon commented on src/main/java/org/dave/compactmachines3/gui/framework/widgets/Widget.java:
```diff
@@ -234,7 +231,7 @@ public void drawBeforeShift(GuiScreen screen) {
      */
     public void draw(GuiScreen screen) {
-        //Logz.debug("Drawing widget: %s, x=%d, y=%d, width=%d, height=%d", this, layoutResult.getX(), layoutResult.getY(), layoutResult.getWidth(), layoutResult.getHeight());
+        //Logz.debug("Drawing widget: {}, x={}, y={}, width={}, height={}", this, layoutResult.getX(), layoutResult.getY(), layoutResult.getWidth(), layoutResult.getHeight());
```
Needs removed instead of commented out

robotgryphon commented on src/main/java/org/dave/compactmachines3/gui/framework/widgets/WidgetPanel.java:
```diff
@@ -47,19 +46,17 @@ public WidgetPanel() {
             int innerX = event.x - widget.getActualX();
             int innerY = event.y - widget.getActualY();
-            //Logz.info("%s: Click: screen@%d,%d, panel@%d,%d", this.toString(), event.x, event.y, innerX, innerY);
+            //Logz.info("{}: Click: screen@{},{}, panel@{},{}", this.toString(), event.x, event.y, innerX, innerY);
```
Needs removed instead of commented out

robotgryphon commented on src/main/java/org/dave/compactmachines3/misc/PlayerEventHandler.java:
```diff
@@ -166,29 +167,29 @@ public static void onPlayerTick(TickEvent.PlayerTickEvent event) {
 
-        if(!ConfigurationHandler.MachineSettings.keepPlayersInside) {
+        if (enteredMachineId != -1 && lastId != enteredMachineId && ConfigurationHandler.MachineSettings.allowEnteringWithoutPSD) {
+            TeleportationTools.addMachineIdToHistory(enteredMachineId, event.player);
```
Solves major issues with other mod teleportation (#408 in particular)

robotgryphon commented on src/main/java/org/dave/compactmachines3/misc/PlayerEventHandler.java:
```diff
@@ -166,29 +167,29 @@ public static void onPlayerTick(TickEvent.PlayerTickEvent event) {
         }
 
-        if(event.player.isCreative() && event.player.canUseCommand(2, "")) {
+        if((event.player.isCreative() || event.player.isSpectator()) && event.player.canUseCommand(2, "")) {
```
Resolves #413

robotgryphon commented on src/main/java/org/dave/compactmachines3/tile/TileEntityFieldProjector.java:
```diff
@@ -368,7 +363,7 @@ public void update() {
             for(BlockPos pos : getInsideBlocks()) {
                 world.setBlockToAir(pos);
-                ((WorldServer)world).spawnParticle(EnumParticleTypes.SMOKE_LARGE, pos.getX(), pos.getY(), pos.getZ(), 10, 0.5D, 0.5D, 0.5D, 0.01D, new int[0]);
+                ((WorldServer)world).spawnParticle(EnumParticleTypes.SMOKE_LARGE, pos.getX(), pos.getY(), pos.getZ(), 10, 0.5D, 0.5D, 0.5D, 0.01D);
```
Why the removal of the int array here?

SizableShrimp replied:
The method was obviously generated by autocomplete. It's a varargs array and omitting it is more idiomatic.

robotgryphon commented on src/main/java/org/dave/compactmachines3/world/WorldSavedDataMachines.java:
```diff
@@ -1,131 +1,198 @@
     }
 
-    @SubscribeEvent
+    @SubscribeEvent(priority = EventPriority.HIGHEST)
```
Why the priority increase?

SizableShrimp replied:
Not sure, probably just to load the WSDM as soon as possible so as to avoid other race conditions. Probably an unnecessary change.

SizableShrimp committed:
Remove unused logs
Simplify one bit of code in `PlayerEventHandler`
